packages/protocol/test-sol/governance/validators/IntergerSortedLinkedList.t.sol

```diff
@@ -0,0 +1,228 @@
+
+    (uint256[] memory keys, uint256[] memory values) = integerSortedLinkedListMock.getElements();
+    assertEq(keys.length, 0);
+    assertEq(values.length, 0);
+  }
+
+  function test_ShouldDecrementNumElements() public {
+    integerSortedLinkedListMock.remove(key);
+    assertEq(integerSortedLinkedListMock.getNumElements(), 0);
+  }
+
+  function test_ShouldUpdateHead() public {
+    integerSortedLinkedListMock.remove(key);
+    assertEq(integerSortedLinkedListMock.head(), 0);
+  }
+
+  function test_ShouldUpdateTail() public {
+    integerSortedLinkedListMock.remove(key);
+    assertEq(integerSortedLinkedListMock.tail(), 0);
+  }
+
+  function test_Reverts_IfKeyIsNotInList() public {
+    vm.expectRevert("key not in list");
+    integerSortedLinkedListMock.remove(key + 1);
+  }
+}
+
+contract IntegerSortedLinkedListTest_PopN is IntegerSortedLinkedListTest {
+  uint256 n = 3;
+  uint256 numElements = 10;
+
+  function setUp() public {
+    super.setUp();
+
+    for (uint256 key = 1; key < numElements + 1; key++) {
+      uint256 value = key * 10;
+      integerSortedLinkedListMock.insert(key, value, key - 1, 0);
+    }
+  }
+
+  function test_ShouldRemoveTheNKeyWithLargestValues() public {
+    integerSortedLinkedListMock.popN(n);
+    for (uint256 key = 1; key < numElements + 1; key++) {
+      if (key <= numElements - n) {
+        assertTrue(integerSortedLinkedListMock.contains(key));
+      } else {
+        assertFalse(integerSortedLinkedListMock.contains(key));
+      }
+    }
+    (uint256[] memory keys, uint256[] memory values) = integerSortedLinkedListMock.getElements();
+    assertEq(keys.length, numElements - n);
+    assertEq(values.length, numElements - n);
+  }
+
+  function test_ShouldReturnTheNKeyWithTheLargestValues() public {
+    uint256[] memory popped = integerSortedLinkedListMock.popN(n);
+
+    uint256[] memory expectedPopped = generateExpectedPopped(numElements, n);
+
+    assertEq(popped, expectedPopped);
+  }
+
+  function test_ShouldDecrementNumElements() public {
+    integerSortedLinkedListMock.popN(n);
+    assertEq(integerSortedLinkedListMock.getNumElements(), numElements - n);
+  }
+
+  function test_ShouldUpdateHead() public {
+    integerSortedLinkedListMock.popN(n);
+    assertEq(integerSortedLinkedListMock.head(), numElements - n);
+  }
+
+  function test_Reverts_IfNIsGreaterThanNumElements() public {
+    vm.expectRevert("not enough elements");
+    integerSortedLinkedListMock.popN(numElements + 1);
+  }
+
+  function generateExpectedPopped(uint256 _numElements, uint256 _n)
+    internal
+    pure
+    returns (uint256[] memory)
+  {
+    require(_n <= _numElements, "Invalid input: _n must be less than or equal to _numElements");
+
+    uint256[] memory expectedPopped = new uint256[](_n);
+    uint256 currentIndex = 0;
+
+    for (uint256 i = _numElements; i > _numElements - _n; i--) {
+      expectedPopped[currentIndex] = i;
+      currentIndex++;
```

Comment:
What about "should maintain invariants when lesser and greater are incorrect" test ?

Reply:
Would prefer not to label it like a test, as this function is used in a test, but not a test itself.